src/ec.rs

```diff
@@ -376,6 +376,32 @@ impl<S> WriterBase<S> {
     }
   }
 
+  #[cfg(any(target_arch = "x86", target_arch = "x86_64"))]
+  pub fn update_cdf_4_sse2(cdf: &mut [u16; 5], val: u32) {
+    #[cfg(target_arch = "x86_64")]
+    use std::arch::x86_64::*;
+    let nsymbs = 4;
+    let rate = 3 + (nsymbs >> 1).min(2) + (cdf[nsymbs] >> 4) as usize;
+    cdf[nsymbs] += 1 - (cdf[nsymbs] >> 5);
+    #[target_feature(enable = "sse2")]
+    unsafe {
+      let val_idx = _mm_set_epi16(0, 0, 0, 0, 3, 2, 1, 0);
+      let val_splat = _mm_set1_epi16(val as i16);
+      let inv_mask = _mm_cmpgt_epi16(val_splat, val_idx);
+      let mask = _mm_xor_si128(inv_mask, _mm_set1_epi16(-1));
+      let v = _mm_loadl_epi64(cdf.as_ptr() as *const __m128i);
+      // *v -= *v >> rate;
+      let shrunk_v = _mm_sub_epi16(v, _mm_srl_epi16(v, _mm_set_epi32(0, 0, 0, rate as i32)));
+      // *v += (32768 - *v) >> rate;
+      let expanded_v = _mm_add_epi16(v, _mm_srl_epi16(
+        _mm_and_si128(_mm_sub_epi16(_mm_set1_epi16(0), v), _mm_set1_epi16(0x7fff)),
```

Comment:
Why can't you just subtract from 32768?

Reply:
32768 is larger than an i16. A way to do this in less than 2 ops is currently escaping me.

Reply:
It fits in u16.

Reply:
yup I'm dumb